### PR TITLE
New device: tuya wifi energy consumption meter

### PR DESCRIPTION
**Pull request by xzeertacat**
Adds support for Tuya Wifi Energy Consumption Meter. Box labelled as V-WIFI-DL02-ES, 
Tuya shows device as "双路电流互感计量器（WiFi版）"

1016076068_max

### CODE REVIEW

File: custom_components/tuya_local/devices/tuya_wifi_energy_consumption_meter.yaml [hunk 1]
```diff
@@ -0,0 +1,157 @@
+name: Tuya Wifi Energy Consumption Meter
```
Comment (make-all) [inline custom_components/tuya_local/devices/tuya_wifi_energy_consumption_meter.yaml:1]: Drop the "Tuya Wifi", as this is redundant for an integration that only supports Tuya devices.  The device name is prepended to all entity names, so should be reasonably short.  Maybe even "Energy meter" is enough.

Also, device and entity names should follow [Home Assistant naming guidelines](https://developers.home-assistant.io/blog/2022/07/10/entity_naming/) in which only the first word is capitalised unless a brand name or abbreviation.  This also applies to the "Total energy" sensors below (but the A and B can be considered abbreviations that should be capitalised).

**Comment by xzeertacat** [comment]
completed modify device name and entity name to follow [Home Assistant naming guidelines](https://developers.home-assistant.io/blog/2022/07/10/entity_naming/). 
thanks.

**Comment by pvinicius27** [comment]
Has anyone managed to connect it to Home Assistant?

**Comment by dkleeman01** [comment]
I have this device and have not been able to get it to connect.   The log file shows:

2024-03-20 14:58:22.721 WARNING (SyncWorker_1) [custom_components.tuya_local.device] Test protocol error 914: Check device key or version
2024-03-20 14:58:37.072 WARNING (SyncWorker_10) [custom_components.tuya_local.device] Test protocol error 914: Check device key or version
2024-03-20 14:58:41.391 WARNING (MainThread) [custom_components.tuya_local.config_flow] Device matches None with quality of 0%. DPS: {"updated_at": 1710971920.4453688, "101": "request", "102": "idle", "103": "monitor", "105": 0, "106": 0, "107": 1243, "110": "normal", "111": 500, "113": "monitor", "115": 0, "116": 0, "117": 1242, "120": "normal", "121": 500, "123": 321, "124": "cloud_net"}

**Comment by make-all** [comment]
It seems like the best match for this device is erz04c_energy_meter

**Comment by dkleeman01** [comment]
Thank you for that.   I'm not sure how to tell tuya-local to use that device profile instead.    I had chosen to install tuya-local because my specific device, "V-WIFI-DL02-ES energy consumption clamp meter"  was itself on the compatibility list.     Nevertheless I am encouraged by your prompt reply.  Thank you.

**Comment by dkleeman01** [comment]
OK I know this is going to break all the rule of how github is supposed to work, but I just want to help the next person coming along this thread.      I was able to get my "V-WIFI-DL02-ES" to work in Home Assistant with TUYA-LOCAL in the following way:

In the folder "homeassistant/config/custom_components/tuya_local/devices", I modified my local copy of: "tuya_energy_meter.yaml" to the following to match the availability of the datapoints (DPS) in the comment above.

When adding the new device in Home Assistant, I needed to specify "protocol: 3.5", as "AUTO" and 3.3, 3.4 did not work.

Here is my modified and working version of tuya_energy_meter.yaml:

name: Dual energy meter
products:
  - id: 79a7z01v3n35kytb
    name: V-WIFI-DL02-ES
primary_entity:
  entity: sensor
  name: Total energy
  class: energy
  dps:
    - id: 123
      name: sensor
      type: integer
      unit: kWh
      class: total_increasing
      mapping:
        - scale: 1000
    - id: 103
      name: Device state A
      type: string
      mapping:
        - dps_val: close
          value: Close
          default: true
        - dps_val: monitor
          value: Monitor
        - dps_val: working
          value: Working
        - dps_val: warning
          value: Warning
    - id: 113
      name: Device state B
      type: string
      mapping:
        - dps_val: close
          value: Close
          default: true
        - dps_val: monitor
          value: Monitor
        - dps_val: working
          value: Working
        - dps_val: warning
          value: Warning	  
    - id: 110
      name: Power Type A
      type: string
      mapping:
        - dps_val: normal
          value: Normal
          default: true
        - dps_val: warn
          value: Warn
    - id: 120
      name: Power Type B
      type: string
      mapping:
        - dps_val: normal
          value: Normal
          default: true
        - dps_val: warn
          value: Warn
    - id: 101
      name: Sync request
      type: string
      mapping:
        - dps_val: idle
          value: Idle
          default: true
        - dps_val: request
          value: Request
    - id: 124
      name: Net state
      type: string
      mapping:
        - dps_val: cloud_net
          value: Cloud
          default: true
        - dps_val: local_net
          value: Local
        - dps_val: no_net
          value: Offline
secondary_entities:
  - entity: sensor
    name: Power A
    category: diagnostic
    class: power
    dps:
      - id: 105
        name: sensor
        type: integer
        unit: W
        force: true
        class: measurement
        mapping:
          - scale: 10
  - entity: sensor
    name: Current A
    category: diagnostic
    class: current
    dps:
      - id: 106
        name: sensor
        type: integer
        unit: A
        class: measurement
        force: true
        mapping:
          - scale: 1000
  - entity: sensor
    name: Voltage A
    category: diagnostic
    class: voltage
    dps:
      - id: 107
        name: sensor
        type: integer
        unit: V
        class: measurement
        force: true
        mapping:
          - scale: 10
  - entity: sensor
    name: Power B
    category: diagnostic
    class: power
    dps:
      - id: 115
        name: sensor
        type: integer
        unit: W
        force: true
        class: measurement
        mapping:
          - scale: 10
  - entity: sensor
    name: Current B
    category: diagnostic
    class: current
    dps:
      - id: 116
        name: sensor
        type: integer
        unit: A
        class: measurement
        force: true
        mapping:
          - scale: 1000
  - entity: sensor
    name: Voltage B
    category: diagnostic
    class: voltage
    dps:
      - id: 117
        name: sensor
        type: integer
        unit: V
        class: measurement
        force: true
        mapping:
          - scale: 10
  - entity: button
    name: Real-time updates
    category: config
    dps:
      - id: 102
        name: button
        type: string
        mapping:
          - dps_val: idle
            value: false
            default: true
          - dps_val: ok
            value: true
        optional: true
  - entity: number
    name: Alarm power setting A
    category: config
    dps:
      - id: 111
        name: value
        type: integer
        unit: W
        range:
          min: 200
          max: 500000
        mapping:
          - step: 100
  - entity: number
    name: Alarm power setting B
    category: config
    dps:
      - id: 121
        name: value
        type: integer
        unit: W
        range:
          min: 200
          max: 500000
        mapping:
          - step: 100

**Comment by make-all** [comment]
Did you try erz04c_energy_meter, which is already a 100% match for the log message you posted above?

And what did you change?  It is too difficult to figure out what might need changing when you just post the whole config like that.

**Comment by dkleeman01** [comment]
I had tried the erz04_c_energy_meter.     It failed to work.   But your suggestion to look at that helped me make the new YAML that worked.

Only when I wrote the YAML to completely match the DPS (and remove all non-matching) did it finally work.    I'm sorry I dont have the bandwidth in life to learn github; I do understand it has a way to show what changed and what stayed the same.   I sure am thankful for your efforts though.   Your hint to me helped make my useless energy meter work on my system.

**Pull request merged**